cpp/src/quantiles/quantiles.cu

```diff
@@ -0,0 +1,286 @@
+#include <stdexcept>
+#include <thrust/detail/execute_with_allocator.h>
+#include <thrust/extrema.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/sort.h>
+#include <thrust/system/cuda/detail/par.h>
+#include <thrust/transform.h>
+#include <rmm/thrust_rmm_allocator.h>
+#include <cudf/column/column_view.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/legacy/reduction.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/sorting.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/bit.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <quantiles/quantiles_util.hpp>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+namespace {
+
+using TScalarResult = double;
+
+struct quantile_index
+{
+    size_type lower;
+    size_type higher;
+    size_type nearest;
+    double fraction;
+
+    quantile_index(size_type count, double quantile)
+    {
+        quantile = std::min(std::max(quantile, 0.0), 1.0);
+
+        double val = quantile * (count - 1);
+        lower = std::floor(val);
+        higher = static_cast<size_t>(std::ceil(val));
+        nearest = static_cast<size_t>(std::nearbyint(val));
+        fraction = val - lower;
+    }
+};
+
+template<typename T>
+std::unique_ptr<scalar>
+select_quantile(T const * begin,
+                size_t size,
+                double quantile,
+                interpolation interpolation)
+{
+    
+    if (size < 2) {
+        auto result_value = get_array_value<TScalarResult>(begin, 0);
+        return std::make_unique<numeric_scalar<TScalarResult>>(result_value);
+    }
+
+    quantile_index idx(size, quantile);
+
+    T a;
+    T b;
+    TScalarResult value{};
+
+    switch (interpolation) {
+    case interpolation::LINEAR:
+        a = get_array_value<T>(begin, idx.lower);
+        b = get_array_value<T>(begin, idx.higher);
+        value = interpolate::linear<TScalarResult>(a, b, idx.fraction);
+        break;
+
+    case interpolation::MIDPOINT:
+        a = get_array_value<T>(begin, idx.lower);
+        b = get_array_value<T>(begin, idx.higher);
+        value = interpolate::midpoint<TScalarResult>(a, b);
+        break;
+
+    case interpolation::LOWER:
+        value = get_array_value<TScalarResult>(begin, idx.lower);
+        break;
+
+    case interpolation::HIGHER:
+        value = get_array_value<TScalarResult>(begin, idx.higher);
+        break;
+
+    case interpolation::NEAREST:
+        value = get_array_value<TScalarResult>(begin, idx.nearest);
+        break;
+
+    default:
+        throw new cudf::logic_error("not implemented");
+    }
```

Comment:
Can you please put this logic back into the `quantile_util.hpp` header. It was being used by `group_quantiles`.

Reply:
well you need to use this on unsupported data?

Reply:
I don't suppose the old version could but it'd be nice if there was the ability to do index indirection. Possibly via a passed function that does that index translation. Then I wouldn't have to generate a sorted column

Reply:
Working on that now.